engine/backends/proverif/proverif_backend.ml

```diff
@@ -112,12 +110,57 @@ module Print = struct
 
   let iblock f = group >> jump 2 0 >> terminate (break 0) >> f >> group
 
+  (* TODO: Give definitions for core / known library functions, cf issues #447, #448 *)
+  let library_functions :
+      (Concrete_ident_generated.name * (AST.expr list -> document)) list =
+    [
+      (Core__ops__try_trait__Try__branch, fun args -> empty);
+      (* just an example *)
+    ]
+
+  let assoc_known_function fname (known_name, _) =
+    Global_ident.eq_name known_name fname
+
+  let translate_known_function fname args =
+    (List.find_exn ~f:(assoc_known_function fname) library_functions |> snd)
+      args
+
+  let is_known_function fname =
+    List.exists ~f:(assoc_known_function fname) library_functions
+
   class print =
     object (print)
       inherit GenericPrint.print as super
       method ty_bool = string "bool"
       method ty_int _ = string "bitstring"
 
+      method! expr' : Generic_printer_base.par_state -> expr' fn =
+        fun ctx e ->
+          let wrap_parens =
+            group
+            >> match ctx with AlreadyPar -> Fn.id | NeedsPar -> iblock braces
+          in
+          match e with
+          | App { f = { e = GlobalVar n; _ }; args } when is_known_function n ->
+              translate_known_function n args
+          (* Desugared `?` operator *)
```

Comment:
Ah, sorry, I meant to make an issue for this... I believe the `Reconstruct_question_marks` phase is ineffective at the moment, because it doesn't recognize the desugared question marks. It may be because of some hoisting in an earlier phase that changes the AST pattern that `Reconstruct_question_marks` is looking for, but I'll go back and see if I can say more precisely what's wrong, or if it is even still an issue or has been resolved in the meantime.

In case it's still there, should I make a new issue of this, or comment on the existing question mark issue #393, because it might be related?

Reply:
Ah... I see! I can help debug that if you need :)
But indeed, things that look for patterns are a bit sensible to rewrite... I think we should always apply them as soon as possible in the list of phases, or we might run into strange things and incompatibilities.


I guess you can file a new issue, #393 is a bit orthogonal to this matter I would say (#393 is more about the `cf_into_monads` phase)

Reply:
But yeah I recently fixed some things about question marks in PR #439, it might help

Reply:
From what I can see now, with the latest changes I still don't get `QuestionMark` nodes. I'll make an issue for it with more context.

The above works for my purposes at the moment, but being able to cleanly match on question marks would of course be nice in the long term.